src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs

```diff
@@ -1209,19 +1209,22 @@ protected virtual async Task<Project> CreateProjectImplAsync(EvaluatedProjectSta
                 foreach (var (newFileName, source) in projectState.Sources)
                 {
                     var documentId = DocumentId.CreateNewId(additionalProjectId, debugName: newFileName);
-                    solution = solution.AddDocument(documentId, newFileName, source, filePath: newFileName);
+                    var (fileName, folders) = GetNameAndFoldersFromPath(newFileName);
+                    solution = solution.AddDocument(documentId, fileName, source, folders: folders, filePath: newFileName);
```

Comment:
❓ I've long suspected our construction of the document info differed slightly from what occurs in practice. Can you provide debugger proof that the update here matches real-world analyzer cases for all four impacted properties?

* `debugName` (part of `DocumentId`)
* `name`
* `folders`
* `filePath`

I believe a correct is required here but want to be certain we fix this for good. 😄

Reply:
I've validated this already for simple cases, but I'll circle back to see how various different edge cases are represented in VS or Omnisharp (specifically files out-of-tree in VS and absolute paths on both) and update the implementation accordingly.

Reply:
Here's the info I've found for VS:

| File location | `debugName` | `name` | `folders` | `filePath` |
|--------------|-----------------|--------|----------|-----------|
| In subdirectory of project included with relative path | absolute path | File name with extension | folder path relative to the project root | absolute path |
| Absolute path to file | Absolute path | file name with extension | empty list | file name and extension | absolute path |
|  Relative path to file outside of project tree | Absolute path | file name with extension | empty list | file name and extension | absolute path |
|  File outside of project tree with `Link` metadata | absolute path | file name with extension | folders from `Link` metadata | original file name and extension (ignores a new name provided in `Link` metadata) | absolute path |

I think the only accurate way to portray this is to extend the `SourceFileCollection` API to also track folder paths and debug name, but since that's an API-breaking change, I don't want to do that without your agreement. Do you think that's the best approach, or should we try to shoehorn in a best-effort solution with the existing API surface?

Reply:
The project root can be determined as:

* If `ProjectState.DefaultPrefix` is not an absolute path, there is no project root.
* If `ProjectState.DefaultPrefix` ends with `/` or `\`, this is the project root.
* Otherwise, `Path.GetDirectoryName(ProjectState.DefaultPrefix)` is the project root.

Reply:
That sounds good to me. I'll update the implementation to do that.